Add functionality to the app

andela-Taiwo opened the pull request:
####What does this PR?
 - Implement the core functionality of the app

#### Description of tasks completed?
- Implement the topics stories
- implement the user stories
- implement the resources

andela-Taiwo committed:
chore(continuous integration) (#1)

add circleci configuration file

[Finishes #158558545]

andela-Taiwo committed:
[#158591672] Set application structure (#2)

* Initial commit

* chore(setup the app)

add app directory
add models directory
add views directory
add controllers directory
add gemfile
add mongod configuration file
[Finishes #158591672]

andela-Taiwo committed:
chore(configuration file) (#3)

add database configuration file
setup the configuration for test environment
add configuration for development environment
add configuration for production environment

[Finishes #158558614]

andela-Taiwo committed:
chore(host app) (#4)

add config.ru to launch the app

[Finishes -#158580023]

andela-Taiwo committed:
chore(add .env) (#5)

add .env file

[Finishes #158580027]

andela-Taiwo committed:
feat(user model) (#6)

- add user model
- add migration for user model

[Finishes #158590842]

andela-Taiwo committed:
[#158627762] Implement user signup (#7)

* feat(user signup)

- add user signup route
- update the user model to accommodate admin status
- add user signup form

[Finishes #158627762]

* fix(signup error)

display the validation errors on signup form
[Finishes ft-create-user-signup-#158627762]

andela-Taiwo committed:
#158590870 Implement user login and logout (#8)

* feat(user signup)

- add user signup route
- update the user model to accomadate admin status
- add user signup form

[Finishes #158627762]

* feat(user signin) implement user signin

add athentication strategy
add login controller
add logout controller

[Finishes #158590870]

* fix(flash messages)

- refactor validation errors
[Finishes #158590870]

andela-Taiwo committed:
feat(topic) (#9)

- add topic model
- add user topics join table

[Finishes #158580149]

andela-Taiwo committed:
feat(resources) (#10)

- add resource model
- add topic_resources join table

[Finishes #158580069]

andela-Taiwo committed:
feat(assets) implement admin can add, edit and delete topic (#11)

add admin dashboard
add topic controller

[Finishes #158764635]

andela-Taiwo committed:
feat(resourses) implement admin can add, edit and delete resource (#12)

add resource form
add resource controllers

[Finishes #158837271]

andela-Taiwo committed:
(user topics) implement user can add topics (#13)

- add view for user dashboard
- add controller for user_topics
- add logic to navigation links
[Finishes #158837252]

andela-Taiwo committed:
(topic_resource) implement admin can add resource to topics (#14)

- add topic_resource controllers
- add view for topic's resources
- add view for a single resource

[Finishes #158558657]

andela-Taiwo committed:
(user-view) Implement user can view resources for a particular topic (#15)

- add user topic-resources controller
- add user topic-resources view
- update README file

[Finishes #158590879]

andela-Taiwo committed:
Revert "(user-view) Implement user can view resources for a particular topic (#15)" (#16)

This reverts commit ce3b40752dd86ad9775531cd0e2c2236b3b60e44.

andela-Taiwo committed:
(user-view) Implement user can view resources for a particular topic (#17)

- add user topic-resources controller
- add user topic-resources view
- update README file

[Finishes #158590879]

andela-Taiwo committed:
(user resources) (#18)

- add logic to return random single resource per topic to user
- fix button colour to portays the right message

[Finishes #158921716]

andela-Taiwo committed:
[#158919173] Implement send resource to users via email (#19)

* (resource)

- add a worker to send daily resource to users
- add email worker to notify enrolled users

[Finishes #158919173]

* (resource)

- fix cronjob

[Finishes #158919173]

andela-Taiwo committed:
(admin) implement admin can manage user (#20)

- add admin user manger view
- add admin user manger controllers
- refactor signup email to user worker
-

[Finishes #158837297]

morriskimani commented on Gemfile:
```diff
@@ -0,0 +1,33 @@
+source 'https://rubygems.org'
+
+# declare the sinatra dependency
+gem 'sinatra'
```
Tip: List gems alphabetically. Rubocop should have flagged this.

morriskimani commented on Gemfile:
```diff
@@ -0,0 +1,33 @@
+gem 'sinatra-activerecord'
+gem 'rake'
+gem 'pg', '~> 0.20.0'
+gem  'rspec'
```
I believe this one, as well as `rack-test` belong in the **test** group.

morriskimani commented on Procfile:
```diff
@@ -0,0 +1,3 @@
+web: bundle exec rackup config.ru -p $PORT
+#worker: bundle exec sidekiq -r ./app/app.rb
```
Let's remove commented out configs. Only have comments when necessary.

morriskimani commented on app/controllers/game.rb:
```diff
@@ -0,0 +1,8 @@
+class Game
```
Is this class being used anywhere in your application?

andela-Taiwo replied:
It was a test eaxmple to integrate the circleCI. I have deleted in my latest branch.

morriskimani commented on Rakefile:
```diff
@@ -0,0 +1,13 @@
+require './app/app'
+require 'sinatra/activerecord/rake'
+
+# bundle exec sidekiq
```
Please remove all commented out code

morriskimani commented on app/config/database.yml:
```diff
@@ -0,0 +1,28 @@
+  database: <%= ENV['DATABASE_NAME'] %>
+
+test:
+  adapter: sqlite3
```
Any reason why you chose to have a different database adapter(sqlite) for your test environment, yet you're using postgresql for all other environments?

morriskimani commented on app/models/user.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+end
+
+
```
Let's get rid of these extra lines. Please make use of a linter

andela-Taiwo replied:
okay.

morriskimani commented on app/models/user.rb:
```diff
@@ -0,0 +1,31 @@
+require_relative 'topic'
+require_relative 'user_topic'
+
+class User < ActiveRecord::Base
```
This model looks good. As a matter of style, however, I'd prefer that you separate the various related items by a blank line. For example, you could separate all these items: **validations**, **associations** and **callbacks**. For a clearer explanation [see this guide](https://github.com/rubocop-hq/rails-style-guide#macro-style-methods)

morriskimani commented on db/migrate/20180706184252_create_users_table.rb:
```diff
@@ -0,0 +1,14 @@
+        t.string   "username"
+        t.string   "email"
+        t.string   "password_digest"
+        t.boolean  "admin", default: false
```
👍 on the default

morriskimani commented on app/controllers/resource_controller.rb:
```diff
@@ -0,0 +1,81 @@
+        end
+
+        app.get '/admin/resource/:id' do
+        check_admin_authentication
```
This doesn't look very sightly. Could you please indent this code appropriately

andela-Taiwo replied:
💯

morriskimani commented on app/controllers/resource_controller.rb:
```diff
@@ -0,0 +1,81 @@
+        app.get '/admin/resource/:id' do
+        check_admin_authentication
+        @resource = Resource.find_by_id(params[:id])
+        if @resource
```
When executing a simple one-line piece of code in a condition, prefer this style:

```ruby
 erb :edit_resource_form if @resource
```

andela-Taiwo replied:
It has an else block.  Just added it in my latest branch.

morriskimani commented on app/controllers/resource_controller.rb:
```diff
@@ -0,0 +1,81 @@
+        end
+
+        app.delete '/admin/resource/:id/delete' do
+          check_admin_authentication
```
This is a bit repetitive. Several route handlers here begin with ` check_admin_authentication`. 

Try making use of a _before filter_ to ensure that this line is executed before all relevant route handlers.

Check the **Filters** section of [Sinatra's documentation](http://sinatrarb.com/intro.html) for a clearer explanation

andela-Taiwo replied:
I have thought of that. Thanks

morriskimani commented on app/helpers/auth_helper.rb:
```diff
@@ -0,0 +1,41 @@
+      end
+
+      def admin_logged_in?
+        true if logged_in? && current_user.admin == true
```
I think this one can be simplified a bit:
- You don't have to explicitly put `true` there, since the result of the expression will be either `true` or `false`
- Also, no need to explicitly check that `current_user.admin == true`.

So, you could simply have this as:
```ruby
def admin_logged_in?
    logged_in? && current_user.admin
end
```

morriskimani commented on app/controllers/resource_controller.rb:
```diff
@@ -0,0 +1,81 @@
+            flash[:success] = 'Successfully add a new resource'
+            redirect '/admin/resource'
+          else
+            @errors = @resource.errors.to_json
```
Do we really have to convert the errors to JSON and then parse the JSON?
Try the snippet below 😄 
```ruby
@resource.errors.messages
```

andela-Taiwo replied:
Okay. I will do that. 
Thanks

morriskimani commented on app/controllers/resource_controller.rb:
```diff
@@ -0,0 +1,81 @@
+
+            @resource.title = params[:title] if params.has_key?('title')
+            @resource.description = params[:description] if params.has_key?('description')
+            @resource.url = params[:url] if params.has_key?('url')
```
Rather than assigning attributes to a resource like this, I'd prefer this approach:
- Have a method, say `update_params`,  that extracts valid update parameters from the `params` hash.
- Inside this route handler, rather than calling `@resource.save`,  you can simply call `@resource.update_attributes(update_params)`

This approach would have 2 advantages:
1. Separation of concerns. Retrieving update params is handled by another method, not the route handler.
2. Rather than assigning attributes one by one, you can perform a mass assignment

andela-Taiwo replied:
Cool.

morriskimani commented on app/controllers/topic_resource_controller.rb:
```diff
@@ -0,0 +1,74 @@
+          erb :topic_resources
+        end
+
+        app.delete '/admin/topic/:id/resource/:resource_id/delete' do
```
Do we need the _delete_ at the end of this route? The HTTP method already specifies that we are using this route to delete a resource.
Is it possible to have this as ` app.delete '/admin/topic/:id/resource/:resource_id`?

morriskimani commented on app/controllers/topic_resource_controller.rb:
```diff
@@ -0,0 +1,74 @@
+          @resource = @topic.resources.find(params[:resource_id])
+          if @topic_resources.delete(@resource)
+            flash[:success] = 'Successfully remove the resource.'
+            redirect '/admin/topic/' + id + '/resources'
```
prefer using ruby string interpolation:
```ruby
 "/admin/topic/#{id}/resources"
```
Also, the same line is repeated in the `else` block that follows. I think you can DRY that out by doing the redirect at the very end of this route handler

morriskimani commented on app/controllers/topic_resource_controller.rb:
```diff
@@ -0,0 +1,74 @@
+          check_admin_authentication
+          @topic = Topic.find(params[:id])
+          id = params[:id]
+          @topic_resources = @topic.resources
```
Is this variable used anywhere?

morriskimani commented on app/controllers/topic_controller.rb:
```diff
@@ -0,0 +1,99 @@
+
+        end
+
+        app.get '/user/topics/:id/resources' do
```
The name of this route suggests that it's purpose is to get all **resources** belonging to a particular **topic**.
However, the view that is rendered `erb: resource` is meant for rendering only a single resource. Is that the intended functionality?

morriskimani commented on app/controllers/resource_controller.rb:
```diff
@@ -0,0 +1,81 @@
+          check_admin_authentication
+          @resource = Resource.find_by_id(params[:id])
+          if @resource
+            erb :edit_resource_form
```
Why do you have this line here? Are you trying to render a form in the middle of an update?

morriskimani commented on app/controllers/topic_resource_controller.rb:
```diff
@@ -0,0 +1,74 @@
+          check_admin_authentication
+          id = params[:id]
+          puts params
+          resource_id = params[:topic][:resource_ids]  if params.has_key?('topic')
```
`resource_ids` might be a better name for this variable since it stores an array of resource_ids.

morriskimani commented on app/controllers/topic_resource_controller.rb:
```diff
@@ -0,0 +1,74 @@
+          end
+        end
+
+        app.post '/admin/topic/:id/resources' do
```
For this route handler, it is clear that you are adding a resource to a topic. However, most of the logic here can be pushed down to the model. As a rule, prefer to push business logic down to the model.

For example, you are validating that an existing topic resource is not added to the topic. Typically, validations are done in the model. 

Just a suggestion, you could have a method in your `Topic` model, say `add_resource`, that handles adding resources to a Topic.

morriskimani commented on app/controllers/user.rb:
```diff
@@ -0,0 +1,115 @@
+
+module  Sinatra
+  module App
+    module SignUp
```
This controller is handling much more than **Sign up**. I suggest that you split it into smaller controllers that handle specific aspects. For example, this can be broken down into:
- Sessions (log in/ log out)
- Admin Dashboard etc.

morriskimani commented on app/controllers/user.rb:
```diff
@@ -0,0 +1,115 @@
+          check_admin_authentication
+          @topics = Topic.order('updated_at  DESC, id  DESC ')
+          @resources = Resource.order('updated_at  DESC, id  DESC ')
+          @users = User.order('updated_at  DESC, id  DESC ')
```
In your admin view that is rendered by this route handler, you are only showing the **count** of these items(topics, resources and users). As such, **retrieving** all items and **ordering** them is not an optimal approach for what you are trying to achieve.

If you only need the count of a resource, you could simply use:
```ruby
@user_count = User.count
```

morriskimani commented on app/controllers/user_topics_controller.rb:
```diff
@@ -0,0 +1,63 @@
+        app.get '/dashboard' do
+          check_authentication
+          user_id = current_user.id
+          @user= User.find(user_id)
```
Do we need to go to the database and retrieve the user? I thought your `current_user` method returns a user, right?

morriskimani reviewed (changes requested):
Good work so far @andela-Taiwo. I've added a few comments here. Some of the issues I've raised occur more than once, please address all instances of such issues. 

Also, some general feedback not captured in the comments:
1. Make use of a linter. You have a lot of unformatted code e.g. unindented blocks of code, ununiform indentation, unnecessary blank lines, commented out code etc.  All these things make the code look messy. Please fix.
2. The installation guide on your Readme lacks some info e.g. how do I set up a local DB?

andela-Taiwo committed:
(user) can view topics (#21)

- add logic to route user to topics view
- add background image to the layout
- add responsiveness to the tables
- refactor the views directory
[Finishes #158590875]

andela-Taiwo committed:
[#159057203] Add tests for the user and topic models (#26)

* (test)
-add test for user model
- add test for topic model
- add test for user controller
[Finishes ##159057203]

* (test)
-add test for user model
- add test for topic model
- add test for user sigup
[Finishes ##159057203]

andela-Taiwo committed:
(test) (#27)

-add test for resource model
[Finishes #159133506]

andela-Taiwo committed:
(test) (#28)

- add test for admin topic controller
- add test for user topic controller
- add topic factories
[Finishes #159139945]

andela-Taiwo committed:
[ #159166308] Add test for resources controller (#29)

* (test)
- add tests for admin resource controller
- add tests for user resource controller
[Finishes #159166308]

* (test)
- update tests for user resource controller
[Finishes #159166308]

andela-Taiwo committed:
(test) (#30)

- add tests for admin user controller
[Finishes ##159191995]

andela-Taiwo committed:
Update the coveralls badge (#31)

andela-Taiwo committed:
[#159133590] Code refactoring (#32)

* ch(code refactor): code refactoring

- add resource controllers for user and admin
- update tests request parameters to use factories
- fix bug the flash messages
[Finishes #159133590]

* (code refactor)

add session controller tests

* (refactor)

- add helper for sending mail
- add test for resource notification helper

* (code refactor)

fix before filter bug
add airbnb rubocop config
fix rubocop violations

* (code refactor)

-fix user session bug on login and signup

* (code refactor)

add redirect to signup controller
add rediret to login controller
add tests for the redirect paths

* (code refactor)

add flash message for invalid user login details
add test for the flash message

* (code refactor)

fix flash error messages

* (code refactor)

fix rubocop violations
